python/lsst/ip/isr/deferredCharge.py

```diff
@@ -0,0 +1,639 @@
+
+        with gainContext(exposure, image, True, gains):
+            for amp in detector.getAmplifiers():
+                ampName = amp.getName()
+
+                ampImage = image[amp.getRawBBox()]
+                if self.config.zeroUnusedPixels:
+                    # We don't apply overscan subtraction, so zero these
+                    # out for now.
+                    ampImage[amp.getRawParallelOverscanBBox()].getArray()[:, :] = 0.0
+                    ampImage[amp.getRawSerialPrescanBBox()].getArray()[:, :] = 0.0
+
+                # The algorithm expects that the readout corner is in
+                # the lower left corner.  Flip it to be so:
+
+                ampData = self.flipData(ampImage.getArray(), amp)
+
+                if ctiCalib.driftScale[ampName] > 0.0:
+                    correctedAmpData = self.local_offset_inverse(ampData,
+                                                                 ctiCalib.driftScale[ampName],
+                                                                 ctiCalib.decayTime[ampName],
+                                                                 self.config.nPixelOffsetCorrection)
+                else:
+                    correctedAmpData = ampData.copy()
+
+                correctedAmpData = self.local_trap_inverse(correctedAmpData,
+                                                           ctiCalib.serialTraps[ampName],
+                                                           ctiCalib.globalCti[ampName],
+                                                           self.config.nPixelTrapCorrection)
+
+                # Undo flips here.  The method is symmetric.
+                correctedAmpData = self.flipData(correctedAmpData, amp)
+                image[amp.getBBox()].getArray()[:, :] = correctedAmpData[:, :]
+
+        return exposure
+
+    @staticmethod
+    def flipData(ampData, amp):
+        """Flip data array such that readout corner is at lower-left.
+
+        Parameters
+        ----------
+        ampData : `np.ndarray`, (nx, ny)
+            Image data to flip.
+        amp : `lsst.afw.cameraGeom.Amplifier`
+            Amplifier to get readout corner information.
+
+        Returns
+        -------
+        ampData : `np.ndarray`, (nx, ny)
+            Flipped image data.
+        """
+        X_FLIP = {ReadoutCorner.LL: False,
+                  ReadoutCorner.LR: True,
+                  ReadoutCorner.UL: False,
+                  ReadoutCorner.UR: True}
+        Y_FLIP = {ReadoutCorner.LL: False,
+                  ReadoutCorner.LR: False,
+                  ReadoutCorner.UL: True,
+                  ReadoutCorner.UR: True}
+
+        if X_FLIP(amp.getReadoutCorner()):
+            ampData = np.fliplr(ampData)
+        if Y_FLIP(amp.getReadoutCorner()):
+            ampData = np.flipud(ampData)
+
+        return ampData
+
+    @staticmethod
+    def local_offset_inverse(inputArr, drift_scale, decay_time, num_previous_pixels=15):
```

Comment:
The default value for `num_previous_pixels` isn't used anywhere, as far as I can see (you supply it everywhere this function is called), and default values like this are potentially harmful, so I'd suggest removing it unless there is good reason to keep it.

Reply:
OK, so reading on, I can see why it is useful to have a default value for something like this, I've just seen the limiting case where this ends up being very bad, and am keen not to see the stack go that way. This is probably fine though, just thought I'd note it though.